sidecar/dns/server.go

```diff
@@ -0,0 +1,422 @@
+	config    Config
+	dnsServer *dns.Server
+}
+
+// Config represents the DNS server configurations.
+type Config struct {
+	DiscoveryClient client.Discovery
+	Port            uint16
+	Domain          string
+}
+
+// NewServer creates a new instance of a DNS server with the given configurations
+func NewServer(config Config) (*Server, error) {
+	err := validate(&config)
+	if err != nil {
+		return nil, err
+	}
+	s := &Server{
+		config: config,
+	}
+
+	// Setup DNS muxing
+	mux := dns.NewServeMux()
+	mux.HandleFunc(config.Domain, s.handleRequest)
+
+	// Setup a DNS server
+	s.dnsServer = &dns.Server{
+		Addr:    fmt.Sprintf(":%d", config.Port),
+		Net:     "udp",
+		Handler: mux,
+	}
+
+	return s, nil
+}
+
+// ListenAndServe starts the DNS server
+func (s *Server) ListenAndServe() error {
+	logrus.Info("Starting DNS server")
+	err := s.dnsServer.ListenAndServe()
+
+	if err != nil {
+		logrus.WithError(err).Errorf("Error starting DNS server")
+	}
+
+	return nil
+}
+
+// Shutdown stops the DNS server
+func (s *Server) Shutdown() error {
+	logrus.Info("Shutting down DNS server")
+	err := s.dnsServer.Shutdown()
+
+	if err != nil {
+		logrus.WithError(err).Errorf("Error shutting down DNS server")
+	} else {
+		logrus.Info("DNS server has shutdown")
+	}
+
+	return err
+}
+
+func (s *Server) handleRequest(w dns.ResponseWriter, request *dns.Msg) {
+	response := new(dns.Msg)
+	response.SetReply(request)
+	response.Extra = request.Extra
+	response.Authoritative = true
+	response.RecursionAvailable = false
+
+	for i, question := range request.Question {
+		err := s.handleQuestion(question, request, response)
+		if err != nil {
+			logrus.WithError(err).Errorf("Error handling DNS question %d: %s", i, question.String())
+			// TODO: what should the dns response return ?
+			break
+		}
+	}
+	err := w.WriteMsg(response)
+	if err != nil {
+		logrus.WithError(err).Errorf("Error writing DNS response")
+	}
+}
+
+func (s *Server) handleQuestion(question dns.Question, request, response *dns.Msg) error {
+
+	switch question.Qclass {
+	case dns.ClassINET:
+	default:
+		response.SetRcode(request, dns.RcodeServerFailure)
+		return fmt.Errorf("unsupported DNS question class: %v", dns.Class(question.Qclass).String())
+	}
+
+	switch question.Qtype {
+	case dns.TypeA:
+	case dns.TypeAAAA:
+	case dns.TypeSRV:
+	default:
+		response.SetRcode(request, dns.RcodeServerFailure)
+		return fmt.Errorf("unsupported DNS question type: %v", dns.Type(question.Qtype).String())
+	}
+
+	serviceInstances, err := s.retrieveServices(question, request, response)
+
+	if err != nil {
+		return err
+	}
+
+	err = s.findMatchingServices(question, request, response, serviceInstances)
+	return err
+
+}
+
+func (s *Server) retrieveServices(question dns.Question, request, response *dns.Msg) ([]*client.ServiceInstance, error) {
+	var ServiceInstances []*client.ServiceInstance
+	var err error
+	// parse query :
+	// Query format:
+	// [tag or endpoint type]*.<service>.service.<domain>.
```

Comment:
Isn't there a better way rather than asking the user to explicitly specify whether they are querying a service or an instance ? I feel that most of the time, the query would be for a service and the instance specific query would be relatively rarer. Maybe the service can be made as default if not specified. And if the instance word is present, then we resolve a specific instance. 

By the way, reserved words should be prefixed with an _ to avoid collision with naturally occurring DNS names.

SRV records - if we are not supporting weights are not being set, they should be set to 0, as per the RFC. 

A records - how are we load balancing the results from the DNS? Are we picking one randomly or using round robin? Can you point me to the function in code where this is being done?